driver-core/src/main/com/mongodb/internal/connection/AbstractMultiServerCluster.java

```diff
@@ -247,46 +246,43 @@ private boolean handleReplicaSetMemberChanged(final ServerDescription newDescrip
-                if (isStalePrimary(newDescription)) {
-                    if (LOGGER.isInfoEnabled()) {
-                        LOGGER.info(format("Invalidating potential primary %s whose (set version, election id) tuple of (%d, %s) "
-                                + "is less than one already seen of (%d, %s)",
-                                newDescription.getAddress(),
-                                setVersion, electionId,
-                                maxSetVersion, maxElectionId));
-                    }
-                    addressToServerTupleMap.get(newDescription.getAddress()).server.resetToConnecting();
+            if (newDescription.getMaxWireVersion() >= SIX_DOT_ZERO_WIRE_VERSION) {
+                if (nullSafeCompareTo(electionId, maxElectionId) > 0
+                        || (nullSafeCompareTo(electionId, maxElectionId) == 0 && nullSafeCompareTo(setVersion, maxSetVersion) >= 0)) {
+                    LOGGER.info(format("Setting max election id to %s and max set version to %d from replica set primary %s",
+                            newDescription.getElectionId(), newDescription.getSetVersion(), newDescription.getAddress()));
+                    maxElectionId = newDescription.getElectionId();
+                    maxSetVersion = newDescription.getSetVersion();
+                } else {
+                    invalidatePotentialPrimary(newDescription);
                     return false;
                 }
-
-                if (!electionId.equals(maxElectionId)) {
-                    if (LOGGER.isInfoEnabled()) {
-                        LOGGER.info(format("Setting max election id to %s from replica set primary %s", electionId,
-                                newDescription.getAddress()));
+            } else {
+                if (setVersion != null && electionId != null) {
+                    if (nullSafeCompareTo(maxSetVersion, setVersion) <= 0
+                            && (nullSafeCompareTo(maxSetVersion, setVersion) != 0 || nullSafeCompareTo(maxElectionId, electionId) <= 0)) {
+                        LOGGER.info(format("Setting max election id to %s from replica set primary %s",
+                                newDescription.getElectionId(), newDescription.getAddress()));
+                        maxElectionId = newDescription.getElectionId();
+                    } else {
+                        invalidatePotentialPrimary(newDescription);
+                        return false;
                     }
-                    maxElectionId = electionId;
                 }
-            }
-
-            if (setVersion != null
-                    && (maxSetVersion == null || setVersion.compareTo(maxSetVersion) > 0)) {
-                if (LOGGER.isInfoEnabled()) {
-                    LOGGER.info(format("Setting max set version to %d from replica set primary %s", setVersion,
-                            newDescription.getAddress()));
+                if (nullSafeCompareTo(setVersion, maxSetVersion) > 0) {
```

Comment:
Note that in the current version of the spec this conditional is nested in `if (setVersion != null && electionId != null)`.  This is an error in the spec that will be corrected soon.  This is the correct spot for it.

CC @nbbeeken

Reply:
Correct and we've merged the fix for this: https://github.com/mongodb/specifications/pull/1312